Randomize sync times

dpsutton opened the pull request:
Randomizes metabase managed sync schedules.

there are two schedules at play: the hourly sync task and the daily field values task.

- when manged by metabase, choose a random minute for the hourly and a random hour for the field values task
- when starting up, look to see if the schedules are manged by metabase, and if the old default (hourly at 50 or 0 (the db had a default of hourly at 50 and the application had a default of hourly at 0)) and if so, randomize them. This was done to not need to figure out how to insert random values during a migration but just do it on application startup. This also allows a spot for more advanced scheduling techniques in the future: ie, take into account how much time each takes and try to schedule them with respect to this.
- put a minute picker on the UI for scheduling

dpsutton committed:
Randomize schedules for dbs when metabase manages them

changes
- allow for minutes in our cron maps so we can choose a random minute
rather than hourly meaning at the top of the hour
- when creating databases, check if user is managing scheduling
otherwise randomize the timings
- on startup, when scheduling, check if the schedules are the old
defaults managed by metabase and randomize them if so

dpsutton committed:
Tests

dpsutton committed:
Frontend display minute picker for hourly schedules

dpsutton committed:
Merge branch 'master' into randomize-sync-times

dpsutton committed:
clean ns for metabase.sync.schedules

dpsutton committed:
Get rid of header comment now that deftests are named

dpsutton committed:
Tests, docstring, and ensure that daily can't revert to hourly

- if just given a map of daily and there are no hours, it turns into
an hourly cronstring and not a daily one

dpsutton committed:
db api tests

sbelak reviewed (approved):
Seems reasonable and a straight improvement of status quo. 

At some point it would be dope if we had probabilistic sync based on underlying change frequency (but this is more a note to self).

dpsutton commented on src/metabase/api/database.clj:
```diff
@@ -572,18 +551,30 @@
+                                                     (cond
+                                                       ;; transition back to metabase managed schedules. the schedule
+                                                       ;; details, even if provided, are ignored
+                                                       (and (get-in database [:details :let-user-control-scheduling])
```
details are the incoming new details. the `database` is the old one currently in the database. It's looking to see if its going back to metabase managed schedules by detecting an edge change of `:let-user-control-schedules` going from true to false.

sbelak replied:
Makes sense. Perhaps name `details` in a way that indicates that so we don't get drive-by changes by people like me who missed that.

dpsutton replied:
i added some more comment. those are destructured bindings so don't have a ton of leeway without introducing a bunch of changes `new-details :details` etc. oh, maybe renaming database to existing-database or something similar can make the point with a smaller footprint of change. good thinking @sbelak

dpsutton committed:
Don't need the seq check since it goes through not-empty

dpsutton committed:
Comment better when checking details in old db vs new values

dpsutton committed:
database -> existing-database for clarity

bigger contrast between existing db values versus new values in the
update db route
